src/bluesearch/database/article.py

```diff
@@ -32,6 +33,50 @@
 from bluesearch.database.identifiers import generate_uid
 
 
+def get_arxiv_id(path: str | Path) -> str | None:
+    """Compute arXiv ID, including version, from file path.
+
+    Parameters
+    ----------
+    path
+        The file path to an arXiv article.
+
+    Returns
+    -------
+    str or None
+        arXiv ID, if possible to compute.
+
+    Raises
+    ------
+    ValueError
+        If no valid arXiv ID could be inferred from the file path.
+
+    References
+    ----------
+    https://arxiv.org/help/arxiv_identifier
+    """
+    path = Path(path)
+
+    # New format, since 2007-04, only needs path stem:
+    # - since 2015-01 have format YYMM.NNNNN (i.e. 5 digits)
+    # - up to 2014-12 have format YYMM.NNNN (i.e. 4 digits)
+    pattern = re.compile(r"\A(\d{4}\.\d{4}\d?v\d+)\Z")
+    match = re.search(pattern, path.stem)
```

Comment:
1. I think for single use regexes the compiling might be redundant as it will be compiled in `re.search` anyway. So I'd suggest writing `re.search(r"\A(\d{4}\.\d{4}\d?v\d+)\Z", path.stem)`.
2. When compiling patterns one can also write `match = pattern.search(path.stem)`.
3. What do `\A` and `\Z` do exactly? The docs say "Matches only at the beginning/end of string", so are they equivalent to "^" and "$"? If we want to match the whole string maybe we should consider using `re.fullmatch(r"\d{4}\.\d{4}\d?v\d+", path.stem)`.

Reply:
1. Yes, you only save the time of a lookup. I mainly used it for a style choice, to separate the pattern definition from the match search. From https://stackoverflow.com/a/452143:
> Python internally compiles and caches regexes whenever you use them anyway (including calls to re.match()), so you're really only changing when the regex gets compiled, and shouldn't be saving much time at all - only the time it takes to check the cache.

2. and 3. — Cool I didn't know about `pattern.search` and `fullmatch`! Changed in ec2ad5f.